// langchain/dist/chains/conversation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConversationChain = void 0;
const llm_chain_1 = require("./llm_chain");
const prompts_1 = require("../prompts");
const memory_1 = require("../memory");
const defaultTemplate = `The following is a friendly conversation between a human and an AI. The AI is talkative and provides lots of specific details from its context. If the AI does not know the answer to a question, it truthfully says it does not know.

Current conversation:
{history}
Human: {input}
AI:`;
const defaultPrompt = new prompts_1.PromptTemplate({
    template: defaultTemplate,
    inputVariables: ["history", "input"],
});
class ConversationChain extends llm_chain_1.LLMChain {
    constructor(fields) {
        var _a, _b, _c;
        super({
            prompt: (_a = fields.prompt) !== null && _a !== void 0 ? _a : defaultPrompt,
            llm: fields.llm,
            outputKey: (_b = fields.outputKey) !== null && _b !== void 0 ? _b : "response",
        });
        this.memory = (_c = fields.memory) !== null && _c !== void 0 ? _c : new memory_1.BufferMemory();
    }
}
exports.ConversationChain = ConversationChain;
