// langchain/dist/cache.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InMemoryCache = exports.BaseCache = exports.getKey = void 0;
const crypto_1 = __importDefault(require("crypto"));
// Takes in an arbitrary number of strings and returns a hash of them
// that can be used as a key in a cache.
const getKey = (...strings) => {
    const hash = crypto_1.default.createHash("sha256");
    strings.forEach((s) => hash.update(s));
    return hash.digest("hex");
};
exports.getKey = getKey;
class BaseCache {
}
exports.BaseCache = BaseCache;
class InMemoryCache extends BaseCache {
    constructor() {
        super();
        Object.defineProperty(this, "cache", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.cache = {};
    }
    lookup(key) {
        return this.cache[key];
    }
    update(key, value) {
        this.cache[key] = value;
    }
}
exports.InMemoryCache = InMemoryCache;
